File: C++/MP3epoc.cpp
#include "findAllFilePaths.h"
#include "getStdOutBufferWidth.h"
#include "processFile.h"
#include "setUpOutputEncoding.h"
#include "shrinkTextWidth.h"
#include "toUpperASCII.h"
#include "version.h"

#include <iostream>
#include <sstream>

#ifdef _WIN32

#define OPTION_PREFIX '/'

#else

#define OPTION_PREFIX ':'

#endif

using namespace MP3epoc;
using namespace std;

namespace
{
    int getConsoleBufferWidth();
    void subMain(int argc, xchar * argv[]);
    void writeError(const xstring & error);
    void writeHelp();
    void
        writeSummary(
        bool wildcardsUsed,
        int processedFileCount,
        int modifiedFileCount
        );

    int getConsoleBufferWidth()
    {
        return getStdOutBufferWidth();
    }

    void subMain(int argc, xchar * argv[])
    {
        setUpOutputEncoding();

        MP3AttributeSet attributeSet;
        xchar formatSpec = XSTR('\0');
        bool optionF = false;

        xstring error;

        RESID errorId;

        vector<const xstring> paths;

        auto
            parseAttrSpec =
            [&attributeSet] (
            const xstring & arg,
            BinaryAttributeStatus status,
            RESID & errorFormatId,
            xchar & badChar)
            {
                auto argLen = arg.length();
                for (unsigned int index = 1; index < argLen; ++index)
                {
                    xchar currentChar = toUpperASCII(arg[index]);

                    MP3Attribute attribute;

                    switch (currentChar)
                    {
                        case XSTR('P'):
                            attribute = MP3Attribute::Private;
                            break;
                        case XSTR('C'):
                            attribute = MP3Attribute::Copyright;
                            break;
                        case XSTR('O'):
                            attribute = MP3Attribute::Original;
                            break;
                        default:
                            errorFormatId = MSG_BAD_ATTRIBUTE;
                            badChar = currentChar;
                            return 0;
                    }
                    if (
                        attributeSet.initAttributeStatus(
                        attribute,
                        static_cast<int>(status))
                        != 0)
                    {
                        errorFormatId = MSG_DOUBLE_ATTRIBUTE;
                        badChar = currentChar;
                        return -1;
                    }
                }
                return 1;
            };

        auto
            parseOpt =
            [&attributeSet, &errorId, &formatSpec, &optionF, argc]
            (const xstring & arg, RESID badOptionErrorId)
            {
                auto argLen = arg.length();
                xchar secondChar = toUpperASCII(arg[1]);
                if (secondChar == XSTR('E'))
                {
                    if (argLen != 3) goto bad_option;

                    EmphasisAttributeStatus emphasisStatus;
                    switch (toUpperASCII(arg[2]))
                    {
                    case XSTR('0'):
                        emphasisStatus = EmphasisAttributeStatus::None;
                        break;
                    case XSTR('1'):
                        emphasisStatus = EmphasisAttributeStatus::e50_15_µs;
                        break;
                    case XSTR('2'):
                        emphasisStatus = EmphasisAttributeStatus::CCITT_j_17;
                        break;
                    case XSTR('X'):
                        emphasisStatus = EmphasisAttributeStatus::Invalid;
                        break;
                    default:
                        goto bad_option;
                    }
                    if (
                        attributeSet.initAttributeStatus(
                        MP3Attribute::Emphasis,
                        static_cast<int>(emphasisStatus))
                        ==
                        0)
                        return 1;
                }
                else
                {
                    if (argLen != 2) goto bad_option;

                    switch (secondChar)
                    {
                    case XSTR('L'):
                    case XSTR('S'):
                        if (formatSpec != XSTR('\0')) break;
                        formatSpec = secondChar;
                        return 1;
                    case XSTR('W'):
                        if (attributeSet.isWholeFile()) break;
                        attributeSet.setWholeFile(true);
                        return 1;
                    case XSTR('F'):
                        if (optionF) break;
                        optionF = true;
                        return 1;
                    case XSTR('?'):
                        if (argc != 1) break;
                        writeHelp();
                        return 0;
                    default:
                        goto bad_option;
                    }
                }
                errorId = MSG_SYNTAX_ERROR;
                return -1;

            bad_option:
                errorId = badOptionErrorId;
                return -1;
            };

        for (int i = 0; i < argc; ++i)
        {
            xstring arg(argv[i]);

            // I guess none of the args may be a null reference.

            auto argLen = arg.length();

            // As a general rule we can state that no argument other than a path
            // may be one character long.

            if (argLen >= 2)
            {
                int firstChar = arg[0];

                switch (firstChar)
                {
                case XSTR('+'):
                    {
                        RESID errorFormatId;
                        xchar badChar;
                        int result =
                            parseAttrSpec(
                            arg,
                            BinaryAttributeStatus::Set,
                            errorFormatId,
                            badChar
                            );
                        if (result > 0) continue;
                        error = getResourceString(errorFormatId, badChar);
                    }
                    goto error;
                case XSTR('-'):
                    {
                        RESID errorFormatId;
                        xchar badChar;
                        int result =
                            parseAttrSpec(
                            arg,
                            BinaryAttributeStatus::NotSet,
                            errorFormatId,
                            badChar
                            );
                        if (result > 0) continue;
                        if (result < 0)
                        {
                            error = getResourceString(errorFormatId, badChar);
                            goto error;
                        }
                    }
                    {
                        int result = parseOpt(arg, MSG_BAD_OPTION_OR_ATTRIBUTE);
                        if (result > 0) continue;
                        if (result < 0) goto error_id;
                    }
                    return;
                case XSTR(OPTION_PREFIX):
                    {
                        int result = parseOpt(arg, MSG_BAD_OPTION);
                        if (result > 0) continue;
                        if (result < 0) goto error_id;
                    }
                    return;
                default:
                    // Must parse arg as a file name.
                    break;
                }
            }

            // Path found.
            paths.push_back(move(arg));
        }

        if (paths.empty())
        {
            errorId = MSG_NO_FILE;
            goto error_id;
        }

        {
            bool anyReadingOption =
                formatSpec != XSTR('\0') ||
                attributeSet.isWholeFile() ||
                optionF;

            if (
                !anyReadingOption &&
                attributeSet.emphasis_().getStatus() ==
                EmphasisAttributeStatus::Invalid)
            {
                errorId = MSG_OPT_EX_IN_WRITING_OP;
                goto error_id;
            }

            vector<const xstring> filePaths;
            int findFilePathsResult =
                findAllFilePaths(writeError, paths, filePaths);
            if (findFilePathsResult < 0) return;

            {
                // Process files.

                MP3AttributeSet attributeSetToApply;

                if (anyReadingOption || attributeSet.isUnspecified())
                {
                    attributeSetToApply = attributeSet.getUnspecified();
                    // If neither of the options /L or /S and no attribute
                    // specification is provided, then MP3epoc lists the
                    // attributes of the specified files as if /L was specified
                    // in the command line.
                    if (formatSpec == XSTR('\0')) formatSpec = XSTR('L');
                }
                else
                {
                    attributeSetToApply = attributeSet;
                    attributeSetToApply.setWholeFile(true);
                    formatSpec = XSTR('\0');
                }

                int processedFileCount = 0;
                int modifiedFileCount = 0;

                MP3GearWheel gearWheel(attributeSetToApply);
                if (!optionF) gearWheel.setKeyFrameNumber(2);

                for (const xstring & filePath: filePaths)
                {
                    ProcessFileResult processFileResult =
                        processFile(
                        filePath,
                        gearWheel,
                        attributeSet,
                        formatSpec
                        );
                    switch (processFileResult)
                    {
                    case ProcessFileResult::Modified:
                        ++modifiedFileCount;
                        // fall through
                    case ProcessFileResult::Unmodified:
                        ++processedFileCount;
                        break;
                    case ProcessFileResult::Unprocessed:
                        break;
                    }
                }

                if (!attributeSetToApply.isUnspecified())
                    writeSummary(
                        findFilePathsResult > 0,
                        processedFileCount,
                        modifiedFileCount
                        );
            }
        }

        return;

    error_id:
        error = getResourceString(errorId);

    error:
        writeError(error);
    }
    
    void writeError(const xstring & error)
    {
        xcerr << error << endl;
    }

    void writeHelp()
    {
        xstring str;
        {
            {
                xostringstream ostream;
                xstring help = getResourceString(MSG_HELP);
                ostream <<
                    XSTR('\n') << APP_NAME << XSTR(' ') << APP_VERSION <<
                    XSTR(" - ") << APP_COPYRIGHT << XSTR(" - ") <<
                    APP_COMPANY << XSTR('\n') << help;
                str = ostream.str();
            }
            {
                int conWidth = getConsoleBufferWidth();
                if (conWidth > 40)
                    str = shrinkTextWidth(str.c_str(), conWidth - 1);
            }
        }
        xcout << str << flush;
    }

    void
        writeSummary(
        bool wildcardsUsed,
        int processedFileCount,
        int modifiedFileCount)
    {
        xstring summary;

        if (wildcardsUsed)
        {
            if (processedFileCount == 0)
                summary = getResourceString(MSG_FILES_PROCESSED_0);
            else
            {
                xstring processedFileString;
                xstring modifiedFileString;

                if (processedFileCount == 1)
                    processedFileString =
                    getResourceString(MSG_FILES_PROCESSED_1);
                else
                    processedFileString =
                    getResourceString(
                    MSG_FILES_PROCESSED_MANY,
                    processedFileCount
                    );

                if (modifiedFileCount == 0)
                    modifiedFileString = getResourceString(MSG_FILES_CHANGED_0);
                else if (modifiedFileCount == 1)
                    modifiedFileString = getResourceString(MSG_FILES_CHANGED_1);
                else
                    modifiedFileString =
                        getResourceString(
                        MSG_FILES_CHANGED_MANY,
                        modifiedFileCount
                        );

                xostringstream stream;
                stream <<
                    processedFileString << XSTR(", ") << modifiedFileString;
                summary = stream.str();
            }
        }
        else
        {
            // An error message should have been printed already; since exactly
            // one file was specified, no summary is required.
            if (processedFileCount == 0) return;

            summary =
                getResourceString(
                modifiedFileCount != 0 ? MSG_FILE_CHANGED : MSG_FILE_NOT_CHANGED
                );
        }

        xcout << summary << XSTR('.') << endl;
    }
}

int xmain(int argc, xchar * argv[])
{
    subMain(argc - 1, argv + 1);
    return 0;
}

File: C#/MP3epoc.cs
using MP3Attr.App;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace MP3Attr.UI
{
    internal static class ConsoleApplication
    {
        private delegate string GetBadOptionDelegate();

        private delegate int ParseAttrSpecDelegate(
            string arg,
            BinaryAttributeStatus status,
            out string errorFormatName,
            out char badChar
            );

        private delegate int ParseOptDelegate(
            string arg,
            GetBadOptionDelegate getBadOption
            );

        private enum ProcessFileResult
        {
            Unprocessed,
            Modified,
            Unmodified,
        }

        private static readonly char OptionPrefix;
        private static readonly string HelpMyMusicDirName;
        private static readonly string HelpNetworkDrivePrefix;

        static ConsoleApplication()
        {
            if (Path.DirectorySeparatorChar != '/')
            {
                OptionPrefix = '/';
                HelpMyMusicDirName = "My Music";
                HelpNetworkDrivePrefix = @"\\";
            }
            else
            {
                OptionPrefix = ':';
                HelpMyMusicDirName = "~/Music";
                HelpNetworkDrivePrefix = "/Volumes/";
            }
        }

        private static int ConsoleBufferWidth
        {
            get
            {
                try
                {
                    bool dummy = Console.CursorVisible;
                }
                catch
                {
                    return -1;
                }
                return Console.BufferWidth;
            }
        }

        private static int FindAllFilePaths(
            Action<string> logError,
            IList<string> paths,
            List<string> filePaths)
        {
            int result = 0;

            foreach (string path in paths)
            {
                string errorFormatName;

                // If any of the specified paths is a directory, provide a
                // useful error description.
                if (Commons.IsDirectory(path))
                    errorFormatName = "PathIsDir";
                else
                {
                    try
                    {
                        bool wildcardsUsed;
                        string[] newFilePaths =
                            Commons.FindFilePaths(path, out wildcardsUsed);
                        if (result >= 0 && wildcardsUsed) ++result;
                        filePaths.AddRange(newFilePaths);
                        continue;
                    }
                    catch (ArgumentException)
                    {
                        errorFormatName = "BadPath";
                    }
                    catch (PathTooLongException)
                    {
                        errorFormatName = "PathTooLong";
                    }
                    catch (UnauthorizedAccessException)
                    {
                        errorFormatName = "AccessDenied";
                    }
                    catch (FileNotFoundException)
                    {
                        errorFormatName = "PathNotFound";
                    }
                }
                result = -1;
                string error =
                    String.Format(
                    Resources.GetString(errorFormatName),
                    Resources.QuotePath(path)
                    );
                logError(error);
            }
            return result;
        }

        private static string FormatResourceWithOptionPrefix(string name)
        {
            return String.Format(Resources.GetString(name), OptionPrefix);
        }

        private static void Main(string[] args)
        {
            SetUpOutputEncoding();

            MP3AttributeSet attributeSet = MP3AttributeSet.Empty;
            char formatSpec = '\0';
            bool optionF = false;

            string error = null;

            string errorName = null;

            IList<string> paths = new List<string>();

            ParseAttrSpecDelegate
                ParseAttrSpec = (
                string arg,
                BinaryAttributeStatus status,
                out string errorFormatName,
                out char badChar
                ) =>
                {
                    int argLen = arg.Length;
                    for (int index = 1; index < argLen; ++index)
                    {
                        char currentChar = char.ToUpperInvariant(arg[index]);

                        MP3Attribute attribute;

                        switch (currentChar)
                        {
                            case 'P':
                                attribute = MP3Attribute.Private;
                                break;
                            case 'C':
                                attribute = MP3Attribute.Copyright;
                                break;
                            case 'O':
                                attribute = MP3Attribute.Original;
                                break;
                            default:
                                errorFormatName = "BadAttribute";
                                badChar = currentChar;
                                return 0;
                        }
                        if (
                            attributeSet.InitAttributeStatus(
                            attribute,
                            (int)status)
                            !=
                            0)
                        {
                            errorFormatName = "DoubleAttribute";
                            badChar = currentChar;
                            return -1;
                        }
                    }
                    errorFormatName = null;
                    badChar = '\0';
                    return 1;
                };

            ParseOptDelegate ParseOpt =
                (string arg, GetBadOptionDelegate getBadOption) =>
                {
                    int argLen = arg.Length;
                    char secondChar = char.ToUpperInvariant(arg[1]);
                    if (secondChar == 'E')
                    {
                        if (argLen != 3) goto bad_option;

                        EmphasisAttributeStatus emphasisStatus;
                        switch (char.ToUpperInvariant(arg[2]))
                        {
                            case '0':
                                emphasisStatus =
                                    EmphasisAttributeStatus.None;
                                break;
                            case '1':
                                emphasisStatus =
                                    EmphasisAttributeStatus.e50_15_µs;
                                break;
                            case '2':
                                emphasisStatus =
                                    EmphasisAttributeStatus.CCITT_j_17;
                                break;
                            case 'X':
                                emphasisStatus =
                                    EmphasisAttributeStatus.Invalid;
                                break;
                            default:
                                goto bad_option;
                        }
                        if (
                            attributeSet.InitAttributeStatus(
                            MP3Attribute.Emphasis,
                            (int)emphasisStatus)
                            ==
                            0)
                            return 1;
                    }
                    else
                    {
                        if (argLen != 2) goto bad_option;

                        switch (secondChar)
                        {
                            case 'L':
                            case 'S':
                                if (formatSpec != '\0') break;
                                formatSpec = secondChar;
                                return 1;
                            case 'W':
                                if (attributeSet.WholeFile) break;
                                attributeSet.WholeFile = true;
                                return 1;
                            case 'F':
                                if (optionF) break;
                                optionF = true;
                                return 1;
                            case '?':
                                if (args.Length != 1) break;
                                WriteHelp();
                                return 0;
                            default:
                                goto bad_option;
                        }
                    }
                    error = Resources.GetString("SyntaxError");
                    return -1;

                bad_option:
                    error = getBadOption();
                    return -1;
                };

            foreach (string arg in args)
            {
                // I guess none of the args may be a null reference.

                int argLen = arg.Length;

                // As a general rule we can state that no argument other than a
                // path may be one character long.

                if (argLen >= 2)
                {
                    int firstChar = arg[0];

                    switch (firstChar)
                    {
                        case '+':
                            {
                                string errorFormatName;
                                char badChar;
                                int result =
                                    ParseAttrSpec(
                                    arg,
                                    BinaryAttributeStatus.Set,
                                    out errorFormatName,
                                    out badChar
                                    );
                                if (result > 0) continue;
                                error =
                                    String.Format(
                                    Resources.GetString(errorFormatName),
                                    badChar
                                    );
                            }
                            goto error;
                        case '-':
                            {
                                string errorFormatName;
                                char badChar;
                                int result =
                                    ParseAttrSpec(
                                    arg,
                                    BinaryAttributeStatus.NotSet,
                                    out errorFormatName,
                                    out badChar
                                    );
                                if (result > 0) continue;
                                if (result < 0)
                                {
                                    error =
                                        String.Format(
                                        Resources.GetString(errorFormatName),
                                        badChar
                                        );
                                    goto error;
                                }
                            }
                            {
                                int result =
                                    ParseOpt(
                                    arg,
                                    () =>
                                    {
                                        return
                                            Resources.GetString(
                                            "BadOptionOrAttribute"
                                            );
                                    }
                                    );
                                if (result > 0) continue;
                                if (result < 0) goto error;
                            }
                            return;
                        default:
                            if (firstChar == OptionPrefix)
                            {
                                int result =
                                    ParseOpt(
                                    arg,
                                    () =>
                                    {
                                        return
                                            FormatResourceWithOptionPrefix(
                                            "BadOption"
                                            );
                                    }
                                    );
                                if (result > 0) continue;
                                if (result < 0) goto error;
                                return;
                            }
                            // Must parse arg as a file name.
                            break;
                    }
                }

                // Path found.
                paths.Add(arg);
            }

            if (paths.Count == 0)
            {
                errorName = "NoFile";
                goto error_name;
            }

            {
                bool anyReadingOption =
                    formatSpec != '\0' || attributeSet.WholeFile || optionF;

                if (
                    !anyReadingOption &&
                    attributeSet.Emphasis.Status ==
                    EmphasisAttributeStatus.Invalid)
                {
                    error = FormatResourceWithOptionPrefix("OptExInWritingOp");
                    goto error;
                }

                List<string> filePaths = new List<string>();
                int findFilePathsResult =
                    FindAllFilePaths(WriteError, paths, filePaths);
                if (findFilePathsResult < 0) return;

                {
                    // Process files.

                    MP3AttributeSet attributeSetToApply;

                    if (anyReadingOption || attributeSet.IsUnspecified)
                    {
                        attributeSetToApply = attributeSet.Unspecified;
                        // If neither of the options /L or /S and no attribute
                        // specification is provided, then MP3epoc lists the
                        // attributes of the specified files as if /L was
                        // specified in the command line.
                        if (formatSpec == '\0') formatSpec = 'L';
                    }
                    else
                    {
                        attributeSetToApply = attributeSet;
                        attributeSetToApply.WholeFile = true;
                        formatSpec = '\0';
                    }

                    int processedFileCount = 0;
                    int modifiedFileCount = 0;

                    MP3GearWheel gearWheel =
                        new MP3GearWheel(attributeSetToApply);
                    if (!optionF) gearWheel.KeyFrameNumber = 2;

                    foreach (string filePath in filePaths)
                    {
                        ProcessFileResult processFileResult =
                            ProcessFile(
                            filePath,
                            gearWheel,
                            attributeSet,
                            formatSpec
                            );
                        if (processFileResult != ProcessFileResult.Unprocessed)
                        {
                            ++processedFileCount;
                            if (processFileResult == ProcessFileResult.Modified)
                                ++modifiedFileCount;
                        }
                    }

                    if (!attributeSetToApply.IsUnspecified)
                        WriteSummary(
                            findFilePathsResult > 0,
                            processedFileCount,
                            modifiedFileCount
                            );
                    }
                }

            return;

        error_name:
            error = Resources.GetString(errorName);

    error:
            WriteError(error);
        }

        private static ProcessFileResult ProcessFile(
            string filePath,
            MP3GearWheel gearWheel,
            MP3AttributeSet attributeSetToView,
            char formatSpec)
        {
            MP3AttributeSet attributeSetBefore;
            try
            {
                // The main difference between invoking gearWheel.ReadAttributes
                // and invoking gearWheel.ApplyAttributes without any attributes
                // to apply is that gearWheel.ReadAttributes throws
                // MP3KeyFrameNotFoundException if the file processed has no key
                // frame, while gearWheel.ApplyAttributes succeeds normally,
                // possibly returning an unspecified or partially unspecified
                // set (which cannot happen if the key frame is found).
                if (gearWheel.AttributeSetToApply.IsUnspecified)
                    attributeSetBefore = gearWheel.ReadAttributes(filePath);
                else
                    attributeSetBefore = gearWheel.ApplyAttributes(filePath);
            }
            catch (Exception e)
            {
                // Errors in file processing are printed to the standard output
                // stream rather than the standard error output stream in order
                // to keep all information in one listing when the output is
                // redirected.
                Console.WriteLine(
                    "{0}: {1}",
                    Resources.GetString("ERROR"),
                    e.Message
                    );
                return ProcessFileResult.Unprocessed;
            }

            if (
                formatSpec != '\0' &&
                attributeSetBefore.Matches(attributeSetToView))
            {
                String format = "{0:" + formatSpec + "}    {1}";
                Console.WriteLine(
                    format,
                    attributeSetBefore,
                    Path.GetFileName(filePath)
                    );
            }
            return
                attributeSetBefore.Matches(gearWheel.AttributeSetToApply) ?
                ProcessFileResult.Unmodified :
                ProcessFileResult.Modified;
        }

        private static void SetUpOutputEncoding()
        {
            Encoding encoding = null;
            if (Console.IsOutputRedirected)
            {
                StreamWriter writer =
                    new StreamWriter(
                        Console.OpenStandardOutput(),
                        encoding = new UTF8Encoding(false)
                        );
                writer.AutoFlush = true;
                Console.SetOut(writer);
            }
            if (Console.IsErrorRedirected)
            {
                StreamWriter writer =
                    new StreamWriter(
                        Console.OpenStandardError(),
                        encoding != null ? encoding : new UTF8Encoding(false)
                        );
                writer.AutoFlush = true;
                Console.SetError(writer);
            }
        }

        private static void WriteError(string error)
        {
            Console.Error.WriteLine(error);
        }

        private static void WriteHelp()
        {
            string help =
                String.Format(
                Resources.GetString("Help"),
                OptionPrefix,
                HelpMyMusicDirName,
                Path.DirectorySeparatorChar,
                HelpNetworkDrivePrefix
                );
            string str =
                Environment.NewLine + AssemblyInfo.Name + ' ' +
                AssemblyInfo.Version + " - " + AssemblyInfo.Copyright + " - " +
                AssemblyInfo.Company + Environment.NewLine + help;
            {
                int width = ConsoleBufferWidth;
                if (width > 40) str = Commons.ShrinkTextWidth(str, width - 1);
            }
            Console.Write(str);
        }

        private static void WriteSummary(
            bool wildcardsUsed,
            int processedFileCount,
            int modifiedFileCount)
        {
            string summary;

            if (wildcardsUsed)
            {
                if (processedFileCount == 0)
                    summary = Resources.GetString("FilesProcessed0");
                else
                {
                    string processedFileString;
                    string modifiedFileString;

                    if (processedFileCount == 1)
                        processedFileString =
                            Resources.GetString("FilesProcessed1");
                    else
                        processedFileString =
                            String.Format(
                            Resources.GetString("FilesProcessedMany"),
                            processedFileCount
                            );

                    if (modifiedFileCount == 0)
                        modifiedFileString =
                            Resources.GetString("FilesChanged0");
                    else if (modifiedFileCount == 1)
                        modifiedFileString =
                            Resources.GetString("FilesChanged1");
                    else
                        modifiedFileString =
                            String.Format(
                            Resources.GetString("FilesChangedMany"),
                            modifiedFileCount
                            );

                    summary =
                        String.Format(
                        "{0}, {1}",
                        processedFileString,
                        modifiedFileString
                        );
                }
            }
            else
            {
                // An error message should have been printed already; since
                // exactly one file was specified, no summary is required.
                if (processedFileCount == 0) return;

                summary =
                    Resources.GetString(
                    modifiedFileCount != 0 ? "FileChanged" : "FileNotChanged"
                    );
            }

            Console.WriteLine("{0}.", summary);
        }
    }
}
